File: presentation/src/main/java/com/github/anrimian/musicplayer/ui/playlist_screens/rename/RenamePlayListPresenter.java
package com.github.anrimian.musicplayer.ui.playlist_screens.rename;

import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.ui.common.error.ErrorCommand;
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser;

import io.reactivex.Scheduler;
import io.reactivex.disposables.CompositeDisposable;
import moxy.InjectViewState;
import moxy.MvpPresenter;

@InjectViewState
public class RenamePlayListPresenter extends MvpPresenter<RenamePlayListView> {

    private final PlayListsInteractor playListsInteractor;
    private final Scheduler uiScheduler;
    private final ErrorParser errorParser;

    private final CompositeDisposable presenterDisposable = new CompositeDisposable();

    private final long playListId;

    private String initialName;

    public RenamePlayListPresenter(long playListId,
                                   PlayListsInteractor playListsInteractor,
                                   Scheduler uiScheduler,
                                   ErrorParser errorParser) {
        this.playListId = playListId;
        this.playListsInteractor = playListsInteractor;
        this.uiScheduler = uiScheduler;
        this.errorParser = errorParser;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().showInputState();
        loadPlayListInfo();
        //compare names and disable apply button, just a little feature
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        presenterDisposable.dispose();
    }

    void onCompleteInputButtonClicked(String playListName) {
        presenterDisposable.add(playListsInteractor.updatePlayListName(playListId, playListName)
                .observeOn(uiScheduler)
                .doOnSubscribe(d -> getViewState().showProgress())
                .subscribe(getViewState()::closeScreen, this::onDefaultError));
    }

    private void onDefaultError(Throwable throwable) {
        ErrorCommand errorCommand = errorParser.parseError(throwable);
        getViewState().showError(errorCommand);
    }

    private void loadPlayListInfo() {
        playListsInteractor.getPlayListObservable(playListId)
                .observeOn(uiScheduler)
                .subscribe(this::onPlayListInfoReceived,
                        this::onDefaultError,
                        getViewState()::closeScreen
                );
    }

    private void onPlayListInfoReceived(PlayList playList) {
        initialName = playList.getName();
        getViewState().showPlayListName(initialName);
    }
}

File: app/src/main/java/com.github.anrimian.musicplayer/ui/playlist_screens/rename/RenamePlayListPresenter.kt
package com.github.anrimian.musicplayer.ui.playlist_screens.rename

import com.github.anrimian.musicplayer.domain.interactors.playlists.PlayListsInteractor
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList
import com.github.anrimian.musicplayer.ui.common.error.parser.ErrorParser
import com.github.anrimian.musicplayer.ui.common.mvp.AppPresenter
import io.reactivex.rxjava3.core.Scheduler

class RenamePlayListPresenter(private val playListId: Long,
                              private val playListsInteractor: PlayListsInteractor,
                              uiScheduler: Scheduler,
                              errorParser: ErrorParser) 
    : AppPresenter<RenamePlayListView>(uiScheduler, errorParser) {
    
    private var initialName: String? = null
    
    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.showInputState()
        loadPlayListInfo()
        //compare names and disable apply button, just a little feature
    }

    fun onCompleteInputButtonClicked(playListName: String) {
        viewState.showProgress()
        playListsInteractor.updatePlayListName(playListId, playListName)
                .subscribeOnUi(viewState::closeScreen, this::onDefaultError)
    }

    private fun onDefaultError(throwable: Throwable) {
        val errorCommand = errorParser.parseError(throwable)
        viewState.showError(errorCommand)
    }

    private fun loadPlayListInfo() {
        playListsInteractor.getPlayListObservable(playListId)
                .subscribeOnUi(this::onPlayListInfoReceived, this::onDefaultError, viewState::closeScreen)
    }

    private fun onPlayListInfoReceived(playList: PlayList) {
        initialName = playList.name
        viewState.showPlayListName(initialName)
    }

}